File: src/input/core/macros.c
#include <assert.h>
#include <stddef.h>   /* NULL */
#include "macros.h"

/*
 * A faithful port of ../../input/macros.py -- button-name sets become buttons_t
 * masks and the macro/step tables are static. Keep the two in lockstep; input/dst.py
 * tier I12 proves they agree under fuzzing.
 */

static buttons_t buttons_at(const macro_t *m, uint32_t elapsed)
{
    uint32_t acc = 0;
    for (uint32_t i = 0; i < m->nsteps; i++) {
        acc += m->steps[i].ticks;
        if (elapsed < acc)
            return m->steps[i].buttons;
    }
    return 0;                                   /* past the end -- caller already guards */
}

void macro_player_init(macro_player_t *p)
{
    assert(p != NULL);
    p->nmacros = 0;
    p->triggers = 0;
    p->prev = 0;
    p->active = -1;
    p->start = 0;
    p->started = false;
    p->last_now = 0;
}

int32_t macro_add(macro_player_t *p, buttons_t trigger)
{
    assert(p != NULL);
    assert(trigger != 0);
    if (p->nmacros >= MACRO_MAX)
        return -1;
    int32_t idx = (int32_t)p->nmacros++;
    macro_t *m = &p->macros[idx];
    m->trigger = trigger;
    m->nsteps = 0;
    m->duration = 0;
    p->triggers |= trigger;
    return idx;
}

int32_t macro_add_step(macro_player_t *p, int32_t idx, buttons_t buttons, uint32_t ticks)
{
    assert(p != NULL);
    assert(idx >= 0 && (uint32_t)idx < p->nmacros);
    assert(ticks >= 1u);
    macro_t *m = &p->macros[idx];
    if (m->nsteps >= MACRO_MAX_STEPS)
        return -1;
    if (m->duration + ticks > MACRO_MAX_TICKS)
        return -1;
    m->steps[m->nsteps].buttons = buttons;
    m->steps[m->nsteps].ticks = ticks;
    m->nsteps++;
    m->duration += ticks;
    return 0;
}

buttons_t macro_step(macro_player_t *p, uint32_t now, buttons_t pressed)
{
    assert(p != NULL);
    assert(!p->started || now >= p->last_now);   /* time must not go backwards */

    buttons_t out = pressed;

    /* a trigger's RISING edge starts its macro, but only when nothing is playing */
    if (p->active < 0) {
        for (uint32_t i = 0; i < p->nmacros; i++) {
            buttons_t trig = p->macros[i].trigger;
            if ((pressed & trig) && !(p->prev & trig)) {
                p->active = (int32_t)i;
                p->start = now;
                break;
            }
        }
    }

    if (p->active >= 0) {
        const macro_t *m = &p->macros[p->active];
        uint32_t elapsed = now - p->start;
        if (elapsed >= m->duration) {            /* sequence finished -> idle */
            p->active = -1;
        } else {
            out &= ~p->triggers;                 /* a trigger press drives the macro, not the button */
            out |= buttons_at(m, elapsed);
        }
    }

    p->prev = pressed;
    p->started = true;
    p->last_now = now;
    /* negative space: playback is bounded -- never active past the macro's duration */
    assert(p->active < 0 || (now - p->start) < p->macros[p->active].duration);
    return out;
}

bool macro_player_playing(const macro_player_t *p)
{
    assert(p != NULL);
    assert(p->active >= -1 && p->active < (int32_t)p->nmacros);
    return p->active >= 0;
}

File: test/reference/input/macros.py
"""
Macros — a post-SOCD output stage that plays a recorded input sequence when a
trigger button is pressed.

A macro binds a trigger button to a timed sequence of button-frames. On the RISING
edge of a trigger, the macro plays once (one-shot): each step drives its button set
for a fixed number of ticks, then the next step, until the sequence ends. While a
macro plays, trigger buttons are removed from the live output (you pressed the
trigger to fire the macro, not to send the raw button) and the active step's
buttons are merged in.

Deliberately minimal and deterministic (keep-it-simple; and DST needs a bounded,
predictable model):
  * ONE macro plays at a time. A trigger that fires while another macro is playing
    is ignored — so playback length is bounded by the longest single macro.
  * Playback is driven by ABSOLUTE time (`now - started_at`), like turbo, so a jump
    in `now` still resolves to the right step.
  * A macro always terminates: after its total duration it goes idle. There is no
    unbounded state (bounded-loop rule) — the negative-space assert enforces it.

Engineering standard: static state (one active index + a start tick), bounded
(step count and duration are capped and asserted), asserts on entry + the negative
space, named limits.
"""
from dataclasses import dataclass

MAX_MACROS = 32                # how many macros one player may hold
MAX_MACRO_STEPS = 64           # per-macro step cap
MAX_MACRO_TICKS = 6000         # per-macro total-duration cap (bounded playback)


@dataclass(frozen=True)
class Macro:
    trigger: str                       # button whose rising edge fires the macro
    steps: tuple                       # ((frozenset(buttons), ticks), ...)

    def __post_init__(self):
        assert self.trigger, "macro needs a trigger button"
        assert 1 <= len(self.steps) <= MAX_MACRO_STEPS, "macro step count out of range"
        assert all(t >= 1 for _, t in self.steps), "every macro step needs ticks >= 1"
        assert 1 <= self.duration <= MAX_MACRO_TICKS, "macro duration out of range"

    @property
    def duration(self):
        return sum(t for _, t in self.steps)

    def buttons_at(self, elapsed):
        """Button set active `elapsed` ticks into playback (caller guards elapsed < duration)."""
        acc = 0
        for buttons, ticks in self.steps:
            acc += ticks
            if elapsed < acc:
                return set(buttons)
        return set()                   # past the end — defensive; caller already guards


def macro(trigger, *steps):
    """Build a Macro from (buttons, ticks) pairs; `buttons` may be any iterable."""
    return Macro(trigger, tuple((frozenset(b), t) for b, t in steps))


class MacroPlayer:
    def __init__(self, macros=()):
        self.macros = tuple(macros)
        assert len(self.macros) <= MAX_MACROS, "too many macros"
        self._triggers = frozenset(m.trigger for m in self.macros)
        self._prev = frozenset()
        self._active = None            # index of the playing macro, or None
        self._start = 0
        self._now = -1

    def step(self, now, pressed):
        """Advance to `now`; return the live set with any active macro merged in."""
        assert now >= self._now, "time must not go backwards"
        pressed = set(pressed)
        out = set(pressed)

        # a trigger's RISING edge starts its macro, but only when nothing is playing
        if self._active is None:
            for i, m in enumerate(self.macros):
                if m.trigger in pressed and m.trigger not in self._prev:
                    self._active = i
                    self._start = now
                    break

        if self._active is not None:
            m = self.macros[self._active]
            elapsed = now - self._start
            if elapsed >= m.duration:            # sequence finished -> idle
                self._active = None
            else:
                out -= self._triggers            # a trigger press drives the macro, not the button
                out |= m.buttons_at(elapsed)

        self._prev = frozenset(pressed)
        self._now = now
        # negative space: playback is bounded — never active past the duration.
        assert self._active is None or 0 <= now - self._start < self.macros[self._active].duration, \
            "macro active past its duration"
        return out
